### PR TITLE
Statistics split into buckets

### LINKED ISSUE #543: Split statistics into multiple buckets
One requirement for Protean's experiments is to be able to get statistics of a given Functional Unit and for that we need to be able to get the measures from only a subset of conodes.

simulation config file:
```toml
Buckets = "a:b c:d-e:f"
```
where a-f are a valid index according to the number of hosts. The syntax is using the Go style of slice indexing and buckets are separated by a space (so we can use `strings.Fields`). Also note the usage of `-` when you need multiple ranges for a single bucket.

The `Buckets` parameter can be specified globally or per simulation if the number of hosts change.

Note: if invalid indices are used, it won't trigger any error but the buckets won't get any statistics.

### PR DESCRIPTION
**Pull request by Gilthoniel**
This PR adds an optional parameter to simulations to define buckets that will be filled with statistics of a subset of the hosts.

Fixes #543

### CODE REVIEW

File: simul/monitor/bucket_stats_test.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
+package monitor
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
```
Comment (nkcr) [inline simul/monitor/bucket_stats_test.go]: This is a rare occasion where true per method unit testing could've been done. 
However those two scenarios capture the essential.

File: simul/monitor/bucket_stats.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+	for _, rule := range rr {
+		if rule.Match(host) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// BucketStats split the statistics into buckets according to network addresses
```
Comment (nkcr) [inline simul/monitor/bucket_stats.go]: splitS

File: simul/monitor/bucket_stats.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+// bucketRule represents a filter that will tell if a measure must
+// be added to a bucket. It follows the logic of a slice indexing
+// so that the lower index is inclusive and the upper one is exclusive.
+type bucketRule struct {
+	low  int
+	high int
+}
+
+// newBucketRule takes a string and parse it to instantiate
```
Comment (nkcr) [inline simul/monitor/bucket_stats.go]: parseS

File: simul/build.go [hunk 4]
```diff
@@ -299,8 +330,13 @@ func mkTestDir() {
 	}
 }
 
-func testFile(name string) string {
-	return "test_data/" + name + ".csv"
+func testFile(name string, index int) string {
```
Comment (nkcr) [inline simul/build.go]: I find the name of this function misleading. Maybe `getTestFile(..)` would be more appropriate?

File: simul/build.go [hunk 5]
```diff
@@ -226,10 +258,9 @@ func RunTest(deployP platform.Platform, rc *platform.RunConfig) (*monitor.Stats,
 		if err != nil {
 			return nil, err
 		}
-		return rs, nil
+		return stats, nil
 	case <-time.After(timeout):
-		monitor.Stop()
-		return rs, errors.New("Simulation timeout")
+		return nil, errors.New("Simulation timeout")
```
Comment (nkcr) [inline simul/build.go]: error strings should not be capitalized

File: simul/build.go [hunk 6]
```diff
@@ -116,25 +117,19 @@ func RunTests(deployP platform.Platform, name string, runconfigs []*platform.Run
...
-	if err != nil {
-		log.Fatal("error opening test file:", err)
-	}
+	files := []*os.File{}
 	defer func() {
-		if err := f.Close(); err != nil {
-			log.Error("Couln't close", f.Name())
+		for _, f := range files {
+			if err := f.Close(); err != nil {
+				log.Error("Couln't close", f.Name())
```
Comment (nkcr) [inline simul/build.go]: I don't know if that applies here, but error strings should not be capitalized

File: simul/README.md [hunk 7]
```diff
@@ -39,6 +39,21 @@ If you use the `onet.SimulationBFTree`, the following variables are also availab
...
+
+- `Buckets` - indices range of the buckets
+
+The parameter is a string where the buckets are separated with spaces and the ranges
+by a dash (e.g. `Buckets = "0:5 5:10-15:20"` that will create a bucket with hosts 0
+to 4 and another one with hosts 5 to 9 and 15 to 19). Range indices can be compared
+to Go slices so that the lower index is inclusive and the higher is exclusive.
+
+A file will be written per bucket and the global one containing the statistics of all
+the conodes will always be present independantly from the parameter. Each file will have
```
Comment (nkcr) [inline simul/README.md]: independently

File: simul/build.go (hunk 6 above)
Comment (Gilthoniel) [reply simul/build.go]: Usually the logs are written as a correct English sentence. You're right about the errors though.

File: simul/build.go (hunk 4 above)
Comment (Gilthoniel) [reply simul/build.go]: Renamed as `generateResultFileName`

**Comment by Gilthoniel** [comment]
@nkcr You're completely right about having comments to the test, we usually try to add those. I simply forgot so thanks for the remainder!

Comments should be fixed.

**Pull request merged**